### ISSUE TITLE
DL3020: False Positive when using variable expansions in ADD

### ISSUE THREAD

**Issue by hhromic** [hadolint/hadolint#769]
- [x] This is a bug report
- [ ] This is a feature request
- [x] I searched existing issues before opening this one

### Expected behavior

Hadolint to not trigger the DL3020 error when using variable expansions in `ADD` instructions when the expansion is valid.

### Actual behavior

Hadolint triggers the DL3020 error when using variable expansions in `ADD` instructions even when the expansion is valid.

### Steps to reproduce the behavior

**Output of `hadolint --version` or
  `docker run --rm hadolint/hadolint hadolint --version` or
  `docker run --rm ghcr.io/hadolint/hadolint hadolint --version`:**

```
$ hadolint --version
Haskell Dockerfile Linter UNKNOWN
```
There seems to be a bug in `--version` as it can be seen. I'm using Hadolint 2.8.0 for Linux downloaded from the Github releases page.

```bash
$ hadolint Dockerfile
Dockerfile.test:3 DL3020 error: Use COPY instead of ADD for files and folders
Dockerfile.test:7 DL3020 error: Use COPY instead of ADD for files and folders
```

**Dockerfile** (if relevant)

```Dockerfile
# hadolint does not seem to expand the variable before evaluating the rules
ARG HADOLINT_BIN=https://github.com/hadolint/hadolint/releases/download/v2.8.0/hadolint-Linux-x86_64
ADD ${HADOLINT_BIN} /usr/local/bin/hadolint

# same for tar-gz archives
ARG ARCHIVE_TGZ=archive.tar.gz
ADD ${ARCHIVE_TGZ} /some/path

# doing the same directly (without variable expansion) does work correctly (no errors reported)
ADD https://github.com/hadolint/hadolint/releases/download/v2.8.0/hadolint-Linux-x86_64 /usr/local/bin/hadolint
ADD archive.tar.gz /some/path
```

**Additional environment details (OS, stack version, etc.)**

Running on Linux (Debian Bullseye).

**Comment by dmp1ce** [comment]
Same issue with DL3006. Are `ARG` variables being expanded at all?

**Comment by thernstig** [comment]
Even the best practices documentation uses variable expansion, making it likely users use it https://docs.docker.com/develop/develop-images/instructions/#add-or-copy:

```Dockerfile
# syntax=docker/dockerfile:1

FROM scratch AS src
ARG DOTNET_VERSION=8.0.0-preview.6.23329.7
ADD --checksum=sha256:270d731bd08040c6a3228115de1f74b91cf441c584139ff8f8f6503447cebdbb \
    https://dotnetcli.azureedge.net/dotnet/Runtime/$DOTNET_VERSION/dotnet-runtime-$DOTNET_VERSION-linux-arm64.tar.gz /dotnet.tar.gz
```

**Comment by jjmaestro** [comment]
Just found this when I was going to report the same issue with `DL3020`. It definitely looks like the variables are not being expanded.

I can understand that there are other issues with variable expansion since those can be set at build time and `hadolint` won't have the values when linting. But ideally, the variables that are set in the Dockerfile should be expanded. A corner-case would be something like

```Dockerfile
ARG FOO
ARG BAR="$FOO"
ADD $FOO /tmp/
```
which I think is a valid Dockerfile but if `FOO` is not set, the result is probably not what was originally intended.

In this case, the linting should probably detect that "the materialized ARG variable is empty" and/or that it depends on other ARG variables not available at the time of linting and it should then skip the error, assuming that everything will work at the time of build.

Similarly, at the very least, `DL3020` should not error when a variable is used.
